### PR TITLE
util: slight performance improvement

### PR DESCRIPTION
**Pull request by xyproto**
* Mainly moving variable declarations out of for loops.
* Benchmark command: `./dev bench pkg/util/encoding --filter=BenchmarkEncodeBytes --verbose --bench-mem --count 5 --timeout 40s`

Before:

```
==================== Test output for //pkg/util/encoding:encoding_test:
initialized metamorphic constant "span-reuse-rate" with value 78
initialized metamorphic constant "span-reuse-rate" with value 91
I220812 23:59:01.159830 1 (gostd) rand.go:176  [-] 1  random seed: 4470005570519829084
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkEncodeBytes
BenchmarkEncodeBytes-12              	88422560	        14.18 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	87415534	        13.90 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	94436716	        13.86 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	87787962	        16.87 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	90102583	        14.24 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending
BenchmarkEncodeBytesDescending-12    	42113778	        26.64 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	47674294	        26.51 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	46883444	        26.37 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	48013256	        30.50 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	45904162	        26.44 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue
BenchmarkEncodeBytesValue-12         	130658758	         9.261 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	126716390	         9.457 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	127298155	         9.422 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	128352916	         9.390 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	127801598	        12.67 ns/op	       0 B/op	       0 allocs/op
PASS
================================================================================
//pkg/util/encoding:encoding_test                                        PASSED in 26.0s

Executed 1 out of 1 test: 1 test passes.
```

After:

```
==================== Test output for //pkg/util/encoding:encoding_test:
initialized metamorphic constant "span-reuse-rate" with value 46
initialized metamorphic constant "span-reuse-rate" with value 89
I220812 23:58:35.860275 1 (gostd) rand.go:176  [-] 1  random seed: -8155315485212279035
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkEncodeBytes
BenchmarkEncodeBytes-12              	87628710	        13.10 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	92259553	        13.23 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	89448531	        13.14 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	90906369	        13.34 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytes-12              	87636229	        13.43 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending
BenchmarkEncodeBytesDescending-12    	40071072	        24.96 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	46545849	        24.79 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	48129660	        27.48 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	49784325	        25.31 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesDescending-12    	46501728	        25.17 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue
BenchmarkEncodeBytesValue-12         	134758836	         8.951 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	133416116	         9.270 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	131264793	         9.155 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	130289151	         9.111 ns/op	       0 B/op	       0 allocs/op
BenchmarkEncodeBytesValue-12         	100000000	        12.60 ns/op	       0 B/op	       0 allocs/op
PASS
================================================================================
//pkg/util/encoding:encoding_test                                        PASSED in 23.7s

Executed 1 out of 1 test: 1 test passes.
```

### CODE REVIEW

**Review by HonoreDB — APPROVED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

**Review by yuzefovich — COMMENTED** [review]
Thanks for your contribution! I only have some nits about the commit message.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86072#-:-N9XBpew-v-gGWOBJY3Z:bi4b7qe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @xyproto)

___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86072#-N9XAGQ2-pfEdUuUlM_R):*
nit: let's improve the commit message to be according to our guidelines (I believe you should be able to access [this](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages) link).

TLDR is to add a short blurb to the "body" of the commit message to explain what the commit is about. Also, for the release note I'd keep `None` but please remove the "category" line - the release note line should be last. If you do want to add a release note about the performance improvement, then the format should be "Release note (performance improvement): <text explaining the improvement and when it's applicable>".

Also, I see that you ran the benchmarks before and after the change, and it'd be good to include the comparison into the commit body. We actually have a useful script in `./scripts/bench` which helps with that.

**Comment by xyproto** [comment]
@yuzefovich Thanks for the feedback!

I tried the `scripts/bench` script, but the results varies a lot based on which branch is given first and last on the command line, so now I'm questioning if the proposed changes has any positive effects on the performance at all.

Here is the output I got, and the first and second output folders in `/tmp` are attached.

```
BENCHES=BenchmarkEncodeBytes PKG=./pkg/util/encoding scripts/bench master perf; BENCHES=BenchmarkEncodeBytes PKG=./pkg/util/encoding scripts/bench perf master
Comparing perf (new) with master (old)

Writing to /tmp/tmp.oT1lucc3yB
Switching to master
+ ./dev bench ./pkg/util/encoding --timeout=5m --filter=BenchmarkEncodeBytes --count=10 --bench-mem -v --stream-output --ignore-cache
Switching to perf
+ ./dev bench ./pkg/util/encoding --timeout=5m --filter=BenchmarkEncodeBytes --count=10 --bench-mem -v --stream-output --ignore-cache
name                      old time/op    new time/op    delta
EncodeBytes-12              14.2ns ± 3%    15.2ns ± 3%  +6.89%  (p=0.000 n=10+10)
EncodeBytesDescending-12    27.5ns ± 2%    28.1ns ± 3%  +2.24%  (p=0.030 n=9+10)
EncodeBytesValue-12         9.76ns ± 3%   10.12ns ± 1%  +3.75%  (p=0.000 n=10+10)

name                      old alloc/op   new alloc/op   delta
EncodeBytes-12               0.00B          0.00B         ~     (all equal)
EncodeBytesDescending-12     0.00B          0.00B         ~     (all equal)
EncodeBytesValue-12          0.00B          0.00B         ~     (all equal)

name                      old allocs/op  new allocs/op  delta
EncodeBytes-12                0.00           0.00         ~     (all equal)
EncodeBytesDescending-12      0.00           0.00         ~     (all equal)
EncodeBytesValue-12           0.00           0.00         ~     (all equal)
Switched to branch 'perf'
Comparing master (new) with perf (old)

Writing to /tmp/tmp.RTv6RBe9fF
Switching to perf
+ ./dev bench ./pkg/util/encoding --timeout=5m --filter=BenchmarkEncodeBytes --count=10 --bench-mem -v --stream-output --ignore-cache
Switching to master
+ ./dev bench ./pkg/util/encoding --timeout=5m --filter=BenchmarkEncodeBytes --count=10 --bench-mem -v --stream-output --ignore-cache
name                      old time/op    new time/op    delta
EncodeBytes-12              15.6ns ± 1%    15.5ns ± 4%   ~     (p=0.780 n=10+10)
EncodeBytesDescending-12    28.4ns ± 4%    28.7ns ± 3%   ~     (p=0.165 n=10+10)
EncodeBytesValue-12         10.1ns ± 1%    10.1ns ± 2%   ~     (p=0.889 n=9+10)

name                      old alloc/op   new alloc/op   delta
EncodeBytes-12               0.00B          0.00B        ~     (all equal)
EncodeBytesDescending-12     0.00B          0.00B        ~     (all equal)
EncodeBytesValue-12          0.00B          0.00B        ~     (all equal)

name                      old allocs/op  new allocs/op  delta
EncodeBytes-12                0.00           0.00        ~     (all equal)
EncodeBytesDescending-12      0.00           0.00        ~     (all equal)
EncodeBytesValue-12           0.00           0.00        ~     (all equal)
Previous HEAD position was 06185e6682 Merge #85925
Switched to branch 'perf'
```

Help in interpreting these results are welcome.

[first.tar.gz](https://github.com/cockroachdb/cockroach/files/9338663/first.tar.gz)
[second.tar.gz](https://github.com/cockroachdb/cockroach/files/9338664/second.tar.gz)

**Comment by yuzefovich** [comment]
Indeed, the results vary quite a bit. I kicked off a run with 50 count on a standalone VM, and here is what I got:
```
name                      old time/op    new time/op    delta
EncodeBytes-24              15.9ns ± 0%    15.9ns ± 0%    ~     (p=0.128 n=50+49)
EncodeBytesDescending-24    30.6ns ± 0%    30.5ns ± 0%  -0.36%  (p=0.000 n=48+48)
EncodeBytesValue-24         11.1ns ± 2%    11.1ns ± 1%  -0.41%  (p=0.000 n=48+50)

name                      old alloc/op   new alloc/op   delta
EncodeBytes-24               0.00B          0.00B         ~     (all equal)
EncodeBytesDescending-24     0.00B          0.00B         ~     (all equal)
EncodeBytesValue-24          0.00B          0.00B         ~     (all equal)

name                      old allocs/op  new allocs/op  delta
EncodeBytes-24                0.00           0.00         ~     (all equal)
EncodeBytesDescending-24      0.00           0.00         ~     (all equal)
EncodeBytesValue-24           0.00           0.00         ~     (all equal)
```
It does appear to be a minor perf improvement overall, so I think we should proceed with this change.

**Comment by xyproto** [comment]
I added the benchmarks to the commit message.

**Review by yuzefovich — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86072)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @xyproto)

___
*[`-- commits` line 3 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86072#-N9XAGQ2-pfEdUuUlM_R):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: let's improve the commit message to be according to our guidelines (I believe you should be able to access [this](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages) link).

TLDR is to add a short blurb to the "body" of the commit message to explain what the commit is about. Also, for the release note I'd keep `None` but please remove the "category" line - the release note line should be last. If you do want to add a release note about the performance improvement, then the format should be "Release note (performance improvement): <text explaining the improvement and when it's applicable>".

Also, I see that you ran the benchmarks before and after the change, and it'd be good to include the comparison into the commit body. We actually have a useful script in `./scripts/bench` which helps with that.
</blockquote></details>

Let's shorten this commit message a bit. We tend to not include instructions for how the benchmark numbers were obtained, only the result of comparison.

Also, the release note is intended to be ultimately read by CockroachDB users, and they don't really want to know "how" performance improvement was achieved, only in what ways it can affect them. In this case actually, I'd not include the release note at all since the difference seems to be very minor. "Release note" also needs to be last in the commit message.

Here is my proposed structure of the commit message:
```
<commit name>

<One sentence about what the commit does.>

<short benchmark comparison>

Release justification: <pick a category>

Release note: None
```

----

I did kick off another longer run of `scripts/bench`:
```
name                      old time/op    new time/op    delta
EncodeBytes-24              15.9ns ± 0%    15.9ns ± 0%  +0.10%  (p=0.000 n=99+100)
EncodeBytesDescending-24    30.6ns ± 0%    30.5ns ± 0%  -0.37%  (p=0.000 n=99+97)
EncodeBytesValue-24         11.1ns ± 2%    11.1ns ± 1%  -0.54%  (p=0.000 n=96+100)

name                      old alloc/op   new alloc/op   delta
EncodeBytes-24               0.00B          0.00B         ~     (all equal)
EncodeBytesDescending-24     0.00B          0.00B         ~     (all equal)
EncodeBytesValue-24          0.00B          0.00B         ~     (all equal)

name                      old allocs/op  new allocs/op  delta
EncodeBytes-24                0.00           0.00         ~     (all equal)
EncodeBytesDescending-24      0.00           0.00         ~     (all equal)
EncodeBytesValue-24           0.00           0.00         ~     (all equal)
```
I'd probably only include the top part in the commit message since the allocations don't change.

**Comment by xyproto** [comment]
This is painfully bureaucratic.
